esp/services/esdl_svc_engine/esdl_monitor.cpp

```diff
@@ -0,0 +1,579 @@
+    limitations under the License.
+############################################################################## */
+
+#include "esdl_monitor.hpp"
+#include "jmd5.hpp"
+#include "esdl_binding.hpp"
+#include "esdl_svc_engine.hpp"
+#include <memory>
+
+class CEsdlInstance : implements IInterface, public CInterface
+{
+public:
+    IMPLEMENT_IINTERFACE;
+    CEsdlInstance() { }
+    CEsdlInstance(const char* defId, IEsdlDefinition* def)
+    {
+        StringBuffer lcid(defId);
+        lcid.trim().toLowerCase();
+        m_defId.set(lcid.str());
+        m_def.set(def);
+    }
+    virtual ~CEsdlInstance()
+    {
+    }
+    StringBuffer m_defId;
+    Owned<IEsdlDefinition> m_def;
+};
+
+class CEsdlShare : implements IEsdlShare, public Thread
+{
+public:
+    IMPLEMENT_IINTERFACE;
+    CEsdlShare() : m_stop(false)
+    {
+    }
+    virtual ~CEsdlShare() {}
+
+    virtual void add(const char* defId, IEsdlDefinition* def) override
+    {
+        if (!defId || !*defId || !def)
+            return;
+        DBGLOG("Adding esdl definition %s to shared cache", defId);
+        Owned<CEsdlInstance> instance = new CEsdlInstance(defId, def);
+        CriticalBlock cb(m_critSect);
+        m_esdlInstanceMap.setValue(defId, instance.getLink());
+    }
+
+    virtual void remove(const char* defId) override
+    {
+        if (!defId || !*defId)
+            return;
+        StringBuffer lcid(defId);
+        lcid.trim().toLowerCase();
+        if (lcid.length() == 0)
+            return;
+        defId = lcid.str();
+        DBGLOG("Removing esdl definition %s from shared cache", defId);
+        {
+            CriticalBlock cb(m_critSect);
+            m_esdlInstanceMap.remove(defId);
+        }
+    }
+
+    virtual Linked<IEsdlDefinition> query(const char* defId) override
+    {
+        if (!defId || !*defId)
+            return nullptr;
+        StringBuffer lcid(defId);
+        lcid.trim().toLowerCase();
+        if (lcid.length() == 0)
+            return nullptr;
+        defId = lcid.str();
+        CEsdlInstance* instance = nullptr;
+        {
+            CriticalBlock cb(m_critSect);
+            Owned<CEsdlInstance>* instancepp = m_esdlInstanceMap.getValue(defId);
+            if (instancepp)
+                instance = *instancepp;
+            if (instance)
+                return instance->m_def.get();
+        }
+        return nullptr;
+    }
+
+    //Thread
+    virtual int run() override
+    {
+        while (!m_stop)
+        {
+            m_waitsem.wait(300000);
+            if(m_stop)
+                break;
+            {
+                CriticalBlock cb(m_critSect);
+                StringArray ids;
+                for (auto& item : m_esdlInstanceMap)
+                {
+                    CEsdlInstance* instance = item.getValue();
+                    if (!instance || !instance->m_def.get())
+                        continue;
+                    CInterface* defptr = dynamic_cast<CInterface*>(instance->m_def.get());
```

Comment:
IsShared (or an equivalent function) should be exposed in the IEsdlDefinition to avoid a dynamic cast.  It is exposing implementation details that could easily break.  E.g., if EspDefinition was changed to

    EsdlDefinition : public CInterfaceOf<IEsdlDefinition>